keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1009 @@
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### [KEP 1495: Volume Populators](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators)
+
+The volume-populators API extension allows you to populate a volume with data from an external data source when the volume is created.
+This is a good solution for restoring a volume from a snapshot or initializing a volume with data from a database backup. However, it does not
+address the desire to use OCI distribution, versioning, and signing for mounted data.
+
+The proposed in-tree OCI VolumeSource provides a direct and integrated approach to mount OCI artifacts, leveraging the existing OCI
+infrastructure for packaging, distribution, and security.
+
+### Custom CSI Plugin
+
+See [https://github.com/warm-metal/container-image-csi-driver](https://github.com/warm-metal/container-image-csi-driver)
+
+An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:
+
+ - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds
+   an additional operational burden. For a generic, vendor-agnostic, and widely-adopted solution this would not make sense.
+ - Supporting the image pull secrets as well as credentials provider will be tricky and needs to be reimplemented with the separate API calls.
+ - External CSI plugins implement their own lifecycle management and garbage collection mechanisms,
```

Comment:
Other considerations:

- kubelet has max parallel image pull constant to maintain the reasonable load on a disk and network. This will not be respected by CSI driver and the only point of integration may be if we move this constant down to runtime
- kubelet has GC logic that is not cleaning up images immidiately in case they will be reused. Also GC logic has it's own thresholds and behavior on eviction. It will be nice to have those integrated
- kubelet exposes metrics on image pulls and we have KEP in place to improve it even further. Having CSI exposing those metrics will require customer to integrate with one more source of data

It is all implementeable in CSI plugin, just adds coplexity for the benefit of being out of tree.

I think the biggest advantage of out of tree will be if it will be super rare use case. But I think a lot of people will choose the image registries as a versioned file distribution mechanism if this feature will be available.
-

Reply:
Thank you, I added the points.